llama-index-core/llama_index/core/node_parser/text/utils.py

```diff
@@ -32,28 +32,29 @@ def split_by_char() -> Callable[[str], List[str]]:
     return lambda text: list(text)
 
 
+def split_by_sentence_tokenizer_internal(text: str, tokenizer) -> List[str]:
+    """Get the spans and then return the sentences.
+
+    Using the start index of each span
+    Instead of using end, use the start of the next span if available
+    """
+    spans = list(tokenizer.span_tokenize(text))
+    sentences = []
+    for i, span in enumerate(spans):
+        start = span[0]
+        if i < len(spans) - 1:
+            end = spans[i + 1][0]
+        else:
+            end = len(text)
+        sentences.append(text[start:end])
+    return sentences
+
+
 def split_by_sentence_tokenizer() -> Callable[[str], List[str]]:
     import nltk
 
     tokenizer = nltk.tokenize.PunktSentenceTokenizer()
-
-    # get the spans and then return the sentences
-    # using the start index of each span
-    # instead of using end, use the start of the next span if available
-    def split(text: str) -> List[str]:
-        spans = list(tokenizer.span_tokenize(text))
-        sentences = []
-        for i, span in enumerate(spans):
-            start = span[0]
-            if i < len(spans) - 1:
-                end = spans[i + 1][0]
-            else:
-                end = len(text)
-            sentences.append(text[start:end])
-
-        return sentences
-
-    return split
+    return lambda text: split_by_sentence_tokenizer_internal(text, tokenizer)
```

Comment:
Does this actually fix the problem? I think you also can't pickle a lambda

Reply:
(I thought, anyways)

Reply:
I change the test file `llama-index-core\tests\ingestion\test_pipeline.py`, edit function `test_pipeline_parallel` like issue's code:

```python
def test_pipeline_parallel() -> None:
    document1 = Document.example()
    document1.id_ = "1"
    document2 = Document(text="One\n\n\nTwo\n\n\nThree.", doc_id="2")

    pipeline = IngestionPipeline(
        transformations=[
            SentenceWindowNodeParser.from_defaults(
                window_size=6,
                window_metadata_key="window",
                original_text_metadata_key="original_text",
            ),
        ],
        docstore=SimpleDocumentStore(),
    )

    num_workers = min(2, cpu_count())
    nodes = pipeline.run(documents=[document1, document2], num_workers=num_workers)
    assert len(nodes) == 20
    assert pipeline.docstore is not None
    assert len(pipeline.docstore.docs) == 2
```

Then run test, it is worked, and ChatGPT told me "In a multiprocessing environment, although lambda functions are generally not easy to serialize using pickle, in this specific case, the tokenizer object captured by the lambda is serializable (since the PunktSentenceTokenizer class from nltk has defined appropriate serialization mechanisms). Therefore, this lambda function can be serialized and passed to other processes as a whole, including its captured environment."

So I submitted PR.

Reply:
I ask ClaudeAI and make some tests like that:
```python
import pickle
import multiprocessing as mp
from typing import List, Callable

def split_by_sentence_tokenizer_internal(text: str, tokenizer) -> List[str]:
    spans = list(tokenizer.span_tokenize(text))
    sentences = []
    for i, span in enumerate(spans):
        start = span[0]
        if i < len(spans) - 1:
            end = spans[i + 1][0]
        else:
            end = len(text)
        sentences.append(text[start:end])
    return sentences

def split_by_sentence_tokenizer() -> Callable[[str], List[str]]:
    import nltk
    tokenizer = nltk.tokenize.PunktSentenceTokenizer()
    return lambda text: split_by_sentence_tokenizer_internal(text, tokenizer)

def test_function(func):
    return func("This is a test sentence. And another one.")

def test_function_reference(func_ref):
    return func_ref()("This is a test sentence. And another one.")

if __name__ == '__main__':
    # Get the function and function reference
    sentence_tokenizer = split_by_sentence_tokenizer()
    sentence_tokenizer_ref = split_by_sentence_tokenizer

    print("Direct Pickling Tests:")
    # Test 1: Direct pickling
    try:
        pickle.dumps(sentence_tokenizer)
        print("Pickling function succeeded")
    except Exception as e:
        print(f"Pickling function failed: {e}")

    try:
        pickle.dumps(sentence_tokenizer_ref)
        print("Pickling function reference succeeded")
    except Exception as e:
        print(f"Pickling function reference failed: {e}")

    print("\nMultiprocessing Tests:")
    # Test 2: Multiprocessing
    with mp.Pool(1) as pool:
        # Test function
        try:
            result = pool.apply(test_function, (sentence_tokenizer,))
            print(f"Multiprocessing with function succeeded: {result}")
        except Exception as e:
            print(f"Multiprocessing with function failed: {e}")

        # Test function reference
        try:
            result = pool.apply(test_function_reference, (sentence_tokenizer_ref,))
            print(f"Multiprocessing with function reference succeeded: {result}")
        except Exception as e:
            print(f"Multiprocessing with function reference failed: {e}")

    print("\nAll tests completed.")
```
result:
```
Direct Pickling Tests:
Pickling function failed: Can't pickle local object 'split_by_sentence_tokenizer.<locals>.<lambda>'
Pickling function reference succeeded

Multiprocessing Tests:
Multiprocessing with function failed: Can't pickle local object 'split_by_sentence_tokenizer.<locals>.<lambda>'
Multiprocessing with function reference succeeded: ['This is a test sentence. ', 'And another one.']

All tests completed.
```

It is work because it is ref?

Reply:
hmmm weird. Well if it works it works